src/crimson/os/seastore/btree/btree_types.h

```diff
@@ -0,0 +1,238 @@
+  extent_ref_count_t refcount = 0; ///< refcount
+  uint32_t checksum = 0; ///< checksum of original block written at paddr (TODO)
+
+  lba_map_val_t() = default;
+  lba_map_val_t(
+    extent_len_t len,
+    pladdr_t pladdr,
+    extent_ref_count_t refcount,
+    uint32_t checksum)
+    : len(len), pladdr(pladdr), refcount(refcount), checksum(checksum) {}
+  bool operator==(const lba_map_val_t&) const = default;
+};
+
+std::ostream& operator<<(std::ostream& out, const lba_map_val_t&);
+
+/**
+ * lba_map_val_le_t
+ *
+ * On disk layout for lba_map_val_t.
+ */
+struct __attribute__((packed)) lba_map_val_le_t {
+  extent_len_le_t len = init_extent_len_le(0);
+  pladdr_le_t pladdr;
+  extent_ref_count_le_t refcount{0};
+  ceph_le32 checksum{0};
+
+  lba_map_val_le_t() = default;
+  lba_map_val_le_t(const lba_map_val_le_t &) = default;
+  explicit lba_map_val_le_t(const lba_map_val_t &val)
+    : len(init_extent_len_le(val.len)),
+      pladdr(pladdr_le_t(val.pladdr)),
+      refcount(val.refcount),
+      checksum(val.checksum) {}
+
+  operator lba_map_val_t() const {
+    return lba_map_val_t{ len, pladdr, refcount, checksum };
+  }
+};
+
+} // namespace lba_manager::btree
+
+namespace backref {
+
+struct backref_map_val_t {
+  extent_len_t len = 0;	///< length of extents
+  laddr_t laddr = L_ADDR_MIN; ///< logical address of extents
+  extent_types_t type = extent_types_t::ROOT;
+
+  backref_map_val_t() = default;
+  backref_map_val_t(
+    extent_len_t len,
+    laddr_t laddr,
+    extent_types_t type)
+    : len(len), laddr(laddr), type(type) {}
+
+  bool operator==(const backref_map_val_t& rhs) const noexcept {
+    return len == rhs.len && laddr == rhs.laddr;
+  }
+};
+
+std::ostream& operator<<(std::ostream &out, const backref_map_val_t& val);
+
+struct __attribute__((packed)) backref_map_val_le_t {
+  extent_len_le_t len = init_extent_len_le(0);
+  laddr_le_t laddr = laddr_le_t(L_ADDR_MIN);
+  extent_types_le_t type = 0;
+
+  backref_map_val_le_t() = default;
+  backref_map_val_le_t(const backref_map_val_le_t &) = default;
+  explicit backref_map_val_le_t(const backref_map_val_t &val)
+    : len(init_extent_len_le(val.len)),
+      laddr(val.laddr),
+      type(extent_types_le_t(val.type)) {}
+
+  operator backref_map_val_t() const {
+    return backref_map_val_t{len, laddr, (extent_types_t)type};
+  }
+};
+
+} // namespace backerf
+
+template <typename key_t, typename val_t>
+struct BtreeCursor {
+  op_context_t ctx;
+  CachedExtentRef parent;
+  uint64_t modifications;
+  key_t key;
+  val_t val;
+  uint16_t pos;
+
+  bool is_valid() const;
+
+  std::unique_ptr<BtreeCursor<key_t, val_t>> duplicate() const {
+    return std::make_unique<BtreeCursor<key_t, val_t>>(*this);
+  }
+};
+
+using LBACursor = BtreeCursor<laddr_t, lba_manager::btree::lba_map_val_t>;
+using LBACursorRef = std::unique_ptr<LBACursor>;
+
+using BackrefCursor = BtreeCursor<paddr_t, backref::backref_map_val_t>;
+using BackrefCursorRef = std::unique_ptr<BackrefCursor>;
```

Comment:
Why `LBACursorRef` and `BackrefCursorRef` are necessary ?

They look to be very small and lightweight structures, is it reasonable to avoid allocate them from heap ?

Reply:
The size of `LBACursor` is 74 bytes and `BackrefCursor` is 66 bytes, which are not as small as they look.

Reply:
I see there are considerable malloc and free overhead there.

It's fine deal with them later, though.